File: AssignmentCS/Client/ClientForm.cs
/*
    CE303 Client-server system
    Student ID: 1904535

    This file contains the entry point of the program (Main function).
    It defines thread-safe functions used to interact with GUI.
*/
using System;
using System.Drawing;
using System.Threading;
using System.Windows.Forms;


namespace ClientNamespace {
    public partial class ClientForm : Form {
        // This class contains functions helpful to keep Client class code clean
        // Additionally, separation of GUI form and Client into separate classes
        // helps with unit testing (where GUI is not really needed).


        Client client;

        [STAThread]
        static void Main() {
            Application.EnableVisualStyles();
            Application.SetCompatibleTextRenderingDefault(false);
            Application.Run(new ClientForm());
        }

        public ClientForm() { 
            InitializeComponent();
            this.Text = $"Client (id = ?)";
            listView_log.View = View.Details;
            listView_log.Columns.Add("Type", 50, HorizontalAlignment.Left);
            listView_log.Columns.Add("Time", 70, HorizontalAlignment.Left);
            listView_log.Columns.Add("Event", -2, HorizontalAlignment.Left);

            client = new Client();
            client.setForm(this);

            Thread.CurrentThread.Name = "ClientForm main thread";
        }

        private void button_give_stock_Click(object sender, EventArgs e) {
            if (listBox_traders.SelectedIndex == -1) {
                MessageBox.Show("Select receiver first");
                return;
            }

            int id_to_give = int.Parse(listBox_traders.SelectedItem.ToString());
            client.requestStockTransfer(id_to_give);
        }

        public void updateStockHolderGUI() {
            InvokeIfRequired(() => {
                if (client.has_stock) {
                    label_stock_holder.Text = $"Stock holder = ME";
                    BackColor = Color.FromName("LightBlue");
                    Activate();
                    TopMost = true;
                    button_give_stock.Enabled = true;
                } else {
                    label_stock_holder.Text = $"Stock holder = {client.stock_holder}";
                    BackColor = default(Color);
                    TopMost = false;
                    button_give_stock.Enabled = false;
                }
                listBox_traders.ClearSelected();
            });
        }

        public void addTrader(string name) {
            InvokeIfRequired(() => {
                listBox_traders.Items.Add(name);
                updateCount();
            });
        }

        public void removeTrader(string name) {
            InvokeIfRequired(() => {
                listBox_traders.Items.Remove(name);
                updateCount();
            });
        }

        public void setTraders(string[] names) {
            InvokeIfRequired(() => {
                listBox_traders.Items.Clear();
                listBox_traders.Items.AddRange(names);
                updateCount();
            });
        }

        public void setWindowTitle(string title) {
            InvokeIfRequired(() => this.Text = title);
        }

        public void Log(ListViewItem item) {
            InvokeIfRequired(() => {
                listView_log.Items.Add(item);
                for (int col = 0; col < listView_log.Columns.Count; col++)
                    listView_log.AutoResizeColumn(col, ColumnHeaderAutoResizeStyle.ColumnContent);
                item.EnsureVisible();
            });
        }

        private void updateCount() {
            label_traders.Text = $"Traders (count = {listBox_traders.Items.Count + 1})";
        }

        private void InvokeIfRequired(Action action) {
            // based on:
            //https://stackoverflow.com/questions/2367718/automating-the-invokerequired-code-pattern

            if (InvokeRequired)
                Invoke(action);
            else
                action();
        }
    }
}

File: AssignmentJava/Client/src/main/ClientForm.java
/*
    CE303 Client-server system
    Student ID: 1904535

    This file contains the entry point of the program (Main function).
    It defines thread-safe functions used to interact with GUI.
*/

package main;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.List;

import static javax.swing.SwingUtilities.invokeLater;

public class ClientForm {
    private JLabel label_traders;
    private JPanel panel;

    private JList<String> list_traders;
    private DefaultListModel<String> list_traders_model;

    private JList<String> list_events;
    private DefaultListModel<String> list_events_model;

    private JButton button_give_stock;
    private JLabel label_events;
    private JLabel label_stock_holder;

    // Why client could be a part of the ClientForm?
    // Because that form is always going to have a Client,
    // and client does not necessarily going to have a form.
    // Idk if that's good design practice, maybe the right thing
    // would be to ask someone about it...
    private Client client;


    private JFrame frame;

    public static void main(String[] args) throws Exception {
        ClientForm client_frame = new ClientForm();
    }

    public ClientForm() throws Exception {
        frame = new JFrame("Client (id = ?)");
        frame.setContentPane(panel);
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.pack();
        frame.setVisible(true);

        list_traders_model = new DefaultListModel<>();
        list_traders.setModel(list_traders_model);

        list_events_model = new DefaultListModel<>();
        list_events.setModel(list_events_model);
        button_give_stock.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                if (list_traders.isSelectionEmpty()) {
                    JavaFix.MessageBox_show("Select receiver first");
                    return;
                }

                int id_to_give = Integer.parseInt(list_traders.getSelectedValue());
                client.requestStockTransfer(id_to_give);
            }
        });

        client = new Client();
        client.setForm(this);
    }

    public void updateStockHolderGUI() {
        invokeLater(() -> {
            Color clr = new Color(173,216,230);
            String stock_holder_label = "Stock holder = ME";;
            if (!client.has_stock) {
                stock_holder_label = String.format("Stock holder = %d", client.stock_holder);
                clr = new Color(240,240,240);
            } else {
                frame.toFront();
                frame.requestFocus();
            }
            label_stock_holder.setText(stock_holder_label);
            frame.getContentPane().setBackground(clr);
            frame.setAlwaysOnTop(client.has_stock);
            button_give_stock.setEnabled(client.has_stock);
            list_traders.clearSelection();
        });
    }

    public void addTrader(String name) {
        invokeLater(() -> {
            list_traders_model.addElement(name);
            updateCount();
        });
    }



    public void removeTrader(String name) {
        invokeLater(() -> {
            list_traders_model.removeElement(name);
            updateCount();
        });
    }

    public void setTraders(List<String> names) {
        invokeLater(() -> {
            list_traders_model.clear();
            for (String name : names)
                list_traders_model.addElement(name);
            updateCount();
        });
    }

    public void setWindowTitle(String title) {
        invokeLater(() -> {
            frame.setTitle(title);
        });
    }

    public void Log(String line) {
        invokeLater(() -> {
            list_events_model.addElement(line);
            //for (int col = 0; col < list_events_model.Columns.Count; col++)
            //    listView_log.AutoResizeColumn(col, ColumnHeaderAutoResizeStyle.ColumnContent);
            list_events.ensureIndexIsVisible(list_events_model.size()-1);
        });
    }

    private void updateCount() {
        invokeLater(() -> {
            label_traders.setText("Traders (count = " + Integer.toString(list_traders_model.size() + 1) + ")");
        });
    }

    //https://stackoverflow.com/questions/7229284/refreshing-gui-by-another-thread-in-java-swing
    //https://stackoverflow.com/questions/2186931/java-pass-method-as-parameter/25005082

    /*private void InvokeIfRequired(Runnable runnable) {
        if (SwingUtilities.isEventDispatchThread())
            runnable.run();
        else
            invokeLater(runnable);
    }*/

}
